### PR TITLE
[CBRD-22707] Btree online load : keep leaf page latched for next insert

### PR DESCRIPTION
**Pull request by razvanradulescu**
http://jira.cubrid.org/browse/CBRD-22707

This is an optimization of online index creation.
It tries to maintain the leaf page latch to fill as much keys as possible.

Summary of changes and idea:

An object of new class btree_insert_list is added into insert_helper context. This is used only by online insert in insert mode.
It holds the entire list of pairs (key, OID) constructed from heap file.
This list is sorted prior to actual btree online task starts (actually - a list of pointers to pairs, since for large keys it may be more efficient and saves space).

For non-insert context which still uses online insert functions, the btree_insert_list class has a specialized constructor which can hold only a pointer to a single key,oid pair.

btree_key_online_index_IB_insert_list is just a wrapper on btree_key_online_index_IB_insert and is able to perform inserts on the same page for consecutive keys.

While btree_split_node_and_advance (btree_search_nonleaf_page) executes, the boundaries of encountered non-leaf pages are saved/updated.
When the leaf page is reached, the current key is inserted since we are sure it satisfies the range;
Then, the code advances to new key and it performs several checks:
 - enough space in page (we assume the key does not exists in page: this may trigger a 'false' abort on the last key that would have fit in the page)
 - the new key to be inserted is compared with the boundary values acquired during tree walk
 - calls btree_leaf_is_key_between_min_max : compares with the actual boundaries stored in the leaf page
   There is a special check : if range search result is BTREE_KEY_BIGGER and there is no 'next' leaf page, we can still insert in the current leaf. Similar for BTREE_KEY_SMALLER, 'prev' leaf page.
   This code is inferred from a safety check already included in debug version of btree_search_leaf_page.
 - btree_search_leaf_page: locates the slot id to be inserted into.
When at least one of the checks fails, the function btree_key_online_index_IB_insert_list exits and the normal btree insert operations resumes;
it will re-enter the function, when will attempt to insert from current key in the new found leaf page.

 - before continuing with inserting a new key, we check if there are any other threads waiting for latching the current leaf page; If yes, processing is aborting forcing the bulk insert to restart from root page. 
To avoid overhead of latch check, we check latch only if a "cost" value is reached; the cost is greater for root page, less for leaf page, and increases with node level; the cost increases when a key is inserted in current leaf page, and proportionally, when overflow page is used or new overflow page is allocated.
TODO : releasing the page can be performed even if only another thread of online index is waiting. This can slow down the index creation specially at start. The check may be enhanced to avoid release of latch if only online index creation threads are waiting.

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/base/perf_monitor.c (3 incorrect indentation(s) found)
```
   267 				  "Num_btree_online_inserts_same_page_hold"),
   269 				  "Num_btree_online_inserts_reject_no_more_keys"),
   272 				  "Num_btree_online_inserts_reject_key_not_in_range"),
```
#### File: src/storage/btree.c (50 incorrect indentation(s) found)
```
  1591 						  PAGE_PTR * leaf_page, BTREE_SEARCH_KEY_HELPER * search_key,
  1592 						  bool * restart, void *other_args);
 33162 btree_online_index_dispatcher (THREAD_ENTRY * thread_p, BTID * btid, OID * class_oid, btree_insert_list * insert_list,
 33332 				       PAGE_PTR * leaf_page, BTREE_SEARCH_KEY_HELPER * search_key, bool * restart,
 33333 				       void *other_args)
 33342 
 33347 						     other_args);
 33349 	{
 33350 	  break;
 33351 	}
 33354 	{
 33355 	  assert (insert_list->m_keys_oids.size () == 1);
 33356 	  break;
 33357 	}
 33361 	{
 33362 	  perfmon_inc_stat (thread_p, PSTAT_BT_ONLINE_NUM_INSERTS_SAME_PAGE_HOLD);
 33363 	}
 33366 	{
 33367 	  /* no more keys in list */
 33368 	  perfmon_inc_stat (thread_p, PSTAT_BT_ONLINE_NUM_REJECT_NO_MORE_KEYS);
```
... More than 20 lines are omitted.
#### File: src/storage/btree.h (8 incorrect indentation(s) found)
```
   546   std::vector < key_oid > m_keys_oids;
   547   std::vector < key_oid * >m_sorted_keys_oids;
   554     btree_insert_list ():m_curr_key (NULL), m_curr_oid (NULL), m_curr_pos (0), m_use_sorted_bulk_insert (false)
   558   btree_insert_list (DB_VALUE * key, OID * oid):btree_insert_list ()
   578   size_t add_key (const TP_DOMAIN * key_type, const DB_VALUE * key, const OID & oid);
   580 #undef BITSET_SIZE
   793 					  btree_insert_list * insert_list, int unique, BTREE_OP_PURPOSE purpose,
   794 					  LOG_LSA * undo_nxlsa);
```
#### File: src/transaction/locator_sr.c (6 incorrect indentation(s) found)
```
  7751 		  btree_insert_list one_item_list (key_dbvalue, inst_oid);
  7777 		      btree_insert_list one_item_list (key_dbvalue, inst_oid);
  8457 		      btree_insert_list one_item_list (old_key, oid);
  8510 			  btree_insert_list one_item_list (new_key, oid);
  8536 			  btree_insert_list old_item_list (old_key, oid);
  8546 			  btree_insert_list new_item_list (new_key, oid);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree.h (2 incorrect indentation(s) found)
```
   814 					  btree_insert_list * insert_list, int unique, BTREE_OP_PURPOSE purpose,
   815 					  LOG_LSA * undo_nxlsa);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree.h (2 incorrect indentation(s) found)
```
   815 					  btree_insert_list * insert_list, int unique, BTREE_OP_PURPOSE purpose,
   816 					  LOG_LSA * undo_nxlsa);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree.h (2 incorrect indentation(s) found)
```
   815 					  btree_insert_list * insert_list, int unique, BTREE_OP_PURPOSE purpose,
   816 					  LOG_LSA * undo_nxlsa);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree.c (1 incorrect indentation(s) found)
```
 33521 						      &helper->insert_helper, key_alread_in_page);
```
#### File: src/storage/btree.h (2 incorrect indentation(s) found)
```
   815 					  btree_insert_list * insert_list, int unique, BTREE_OP_PURPOSE purpose,
   816 					  LOG_LSA * undo_nxlsa);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree.c (1 incorrect indentation(s) found)
```
 33521 						      &helper->insert_helper, key_alread_in_page);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree.c (1 incorrect indentation(s) found)
```
 33521 						      &helper->insert_helper, key_alread_in_page);
```
#### File: src/storage/heap_file.c (3 incorrect indentation(s) found)
```
   695 					      int func_col_id, int func_attr_index_start, TP_DOMAIN * midxkey_domain);
 12466 			   int func_attr_index_start, TP_DOMAIN * midxkey_domain)
 12582 			    FUNCTION_INDEX_INFO * func_index_info, TP_DOMAIN * midxkey_domain)
```
#### File: src/storage/heap_file.h (1 incorrect indentation(s) found)
```
   482 					     TP_DOMAIN * midxkey_domain);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/storage/btree.h [hunk 1]
```diff
@@ -532,6 +534,123 @@ struct btree_object_info
...
+#if !defined (NDEBUG)
+  std::vector <advance_page_trace> m_trace;
+  void add_trace (advance_page_trace &&trace_point)
+    {
+      m_trace.push_back (trace_point);
+    }
+#endif
+};
+
+struct btree_insert_list
```
Comment (ribram) [inline src/storage/btree.h]: can we expose here just a class that collects the key_oid's to try to keep header clean? one possibility is to move page_key_boundary to insert_helper or to another structure static to btree.c.

File: src/storage/btree_load.c [hunk 2]
```diff
@@ -5094,9 +5081,19 @@ index_builder_loader_task::execute (cubthread::entry &thread_ref)
...
+  PERF_UTIME_TRACKER time_insert_task = PERF_UTIME_TRACKER_INITIALIZER;
+  PERF_UTIME_TRACKER_START (&thread_ref, &time_insert_task);
+
+  PERF_UTIME_TRACKER time_prepare_task = PERF_UTIME_TRACKER_INITIALIZER;
+  PERF_UTIME_TRACKER_START (&thread_ref, &time_prepare_task);
+
+  m_insert_list.prepare_list ();
+
+  PERF_UTIME_TRACKER_TIME (&thread_ref, &time_prepare_task, PSTAT_BT_ONLINE_PREPARE_TASK);
+
```
Comment (ribram) [inline src/storage/btree_load.c]: shouldn't timer for PSTAT_BT_ONLINE_INSERT_TASK start after prepare? if so, only one tracker is needed.

File: src/storage/btree.c [hunk 3]
```diff
@@ -5074,6 +5091,57 @@ btree_search_nonleaf_page (THREAD_ENTRY * thread_p, BTID_INT * btid, PAGE_PTR pa
 	  *slot_id = middle;
 	  *child_vpid = non_leaf_rec.pnt;
 
+	  if (page_bounds != NULL)
+	    {
+	      if (!page_bounds->m_is_inf_right_key)
```
Comment (ribram) [inline src/storage/btree.c]: can all this be moved to page_bounds member function? the code is repeated for each case (c == 0, < 0 si > 0).

File: src/storage/btree.c [hunk 4]
```diff
@@ -33301,6 +33475,153 @@ btree_online_index_dispatcher (THREAD_ENTRY * thread_p, BTID * btid, DB_VALUE *
...
+	{
+	  /* no more space in page */
+	  perfmon_inc_stat (thread_p, PSTAT_BT_ONLINE_NUM_REJECT_NO_SPACE);
+	  break;
+	}
+
+      /* compare with boundary keys : NULL keys means INF bound, no check is required */
+      if (!insert_list->m_boundaries.m_is_inf_left_key)
+	{
+	  int c;
```
Comment (ribram) [inline src/storage/btree.c]:
```suggestion
	  DB_VALUE_COMPARE_RESULT c;
```

File: src/storage/btree.c [hunk 5]
```diff
@@ -33301,6 +33475,153 @@ btree_online_index_dispatcher (THREAD_ENTRY * thread_p, BTID * btid, DB_VALUE *
...
+	  c = btree_compare_key (&insert_list->m_boundaries.m_left_key, curr_key, btid_int->key_type, 1, 1, NULL);
+	  if (c != DB_LT && c != DB_EQ)
```
Comment (ribram) [inline src/storage/btree.c]: shouldn't be `c != DB_GT`?

File: src/storage/btree.c [hunk 6]
```diff
@@ -33301,6 +33475,153 @@ btree_online_index_dispatcher (THREAD_ENTRY * thread_p, BTID * btid, DB_VALUE *
...
+	  int c;
+	  c = btree_compare_key (curr_key, &insert_list->m_boundaries.m_right_key, btid_int->key_type, 1, 1, NULL);
+	  if (c != DB_LT)
+	    {
+	      perfmon_inc_stat (thread_p, PSTAT_BT_ONLINE_NUM_REJECT_KEY_NOT_IN_RANGE2);
+	      break;
+	    }
+	}
+
+      error_code = btree_leaf_is_key_between_min_max (thread_p, btid_int, *leaf_page, curr_key, search_key);
```
Comment (ribram) [inline src/storage/btree.c]: is this necessary if boundaries are checked?

File: src/storage/btree.c [hunk 7]
```diff
@@ -33301,6 +33475,153 @@ btree_online_index_dispatcher (THREAD_ENTRY * thread_p, BTID * btid, DB_VALUE *
...
+	      perfmon_inc_stat (thread_p, PSTAT_BT_ONLINE_NUM_REJECT_KEY_FALSE_FAILED_RANGE2);
+	    }
+	  else
+	    {
+	      perfmon_inc_stat (thread_p, PSTAT_BT_ONLINE_NUM_REJECT_KEY_NOT_IN_RANGE3);
+	      break;
+	    }
+	}
+
+      error_code = btree_search_leaf_page (thread_p, btid_int, *leaf_page, curr_key, search_key);
```
Comment (ribram) [inline src/storage/btree.c]: this may be optimized in future patch to search only between current key and min/max

File: src/storage/btree.c (hunk 5 above)
Comment (razvanradulescu) [reply src/storage/btree.c]: Code is compare (left_key , curr_key);
the expected order is: left_key  <= curr_key < right_key

File: src/storage/btree_load.c (hunk 2 above)
Comment (razvanradulescu) [reply src/storage/btree_load.c]: insert task time = prepare task time + execution time
execution time is not measure separately
prepare task time = mostly time required for sorting of chunk.

File: src/storage/btree.c (hunk 6 above)
Comment (razvanradulescu) [reply src/storage/btree.c]: a leaf page may have fence keys; the fence keys are not mandatory the same with boundary key from the parent non-leaf page.

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree.c (4 incorrect indentation(s) found)
```
  5016 
 33328 			       OID * oid, int unique, BTREE_OP_PURPOSE purpose, LOG_LSA * undo_nxlsa)
 33349 				    btree_insert_list * insert_list, int unique, BTREE_OP_PURPOSE purpose,
 33350 				    LOG_LSA * undo_nxlsa)
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by ribram** [comment]
please fix your indent hook :)

File: src/storage/heap_file.c [hunk 8]
```diff
@@ -12579,7 +12579,7 @@ heap_midxkey_key_generate (THREAD_ENTRY * thread_p, RECDES * recdes, DB_MIDXKEY
 DB_VALUE *
 heap_attrinfo_generate_key (THREAD_ENTRY * thread_p, int n_atts, int *att_ids, int *atts_prefix_length,
 			    HEAP_CACHE_ATTRINFO * attr_info, RECDES * recdes, DB_VALUE * db_valuep, char *buf,
-			    FUNCTION_INDEX_INFO * func_index_info)
+			    FUNCTION_INDEX_INFO * func_index_info, TP_DOMAIN * midxkey_domain)
```
[the remark this answers is not in the document]
Comment (razvanradulescu) [reply src/storage/heap_file.c]: added some info on function headers;
When generating midxkey values, the domain was not set (was set as NULL).
During sorting, we use generic compare function which for midxkey values, extracts the domain of midxkey component.

File: src/storage/btree_load.c [hunk 9]
```diff
@@ -4920,7 +4920,8 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
 
       /* Generate the key. */
       p_dbvalue = heap_attrinfo_generate_key (thread_p, n_attrs, &attrids[attr_offset], p_prefix_length, attr_info,
-					      &cur_record, &dbvalue, aligned_midxkey_buf, p_func_idx_info);
+					      &cur_record, &dbvalue, aligned_midxkey_buf, p_func_idx_info,
+                                              const_cast <TP_DOMAIN *>(key_type));
```
[the remark this answers is not in the document]
Comment (razvanradulescu) [reply src/storage/btree_load.c]: fixed

File: src/transaction/locator_sr.c [hunk 10]
```diff
@@ -7748,8 +7748,9 @@ locator_add_or_remove_index_internal (THREAD_ENTRY * thread_p, RECDES * recdes,
 	      if (index->index_status == OR_ONLINE_INDEX_BUILDING_IN_PROGRESS)
 		{
 		  /* Online index is currently loading. */
+		  btree_insert_list one_item_list (key_dbvalue, inst_oid);
 		  error_code =
-		    btree_online_index_dispatcher (thread_p, &btid, key_dbvalue, class_oid, inst_oid, unique_pk,
+		    btree_online_index_dispatcher (thread_p, &btid, class_oid, &one_item_list, unique_pk,
```
[the remark this answers is not in the document]
Comment (razvanradulescu) [reply src/transaction/locator_sr.c]: fixed

File: src/storage/btree.h (hunk 1 above)
Comment (razvanradulescu) [reply src/storage/btree.h]: The next best candidate structure would be insert_helper; however this requires an initializer macro, we would need an initializer for DB_VALUE for instance. I don't think we want to keep such constructs in the code.
I think is better to have this change into a future PR where insert_helper is a struct with proper constructor and does not require initializer.

File: src/storage/btree.c (hunk 3 above)
Comment (razvanradulescu) [reply src/storage/btree.c]: moved

File: src/storage/btree.c [hunk 11]
```diff
@@ -35057,3 +35307,284 @@ btree_is_single_object_key (THREAD_ENTRY * thread_p, BTID_INT * btid_int, BTREE_
...
+
+int
+page_key_boundary::update_boundary_eq (THREAD_ENTRY * thread_p, BTID_INT * btid, PAGE_PTR page_ptr,
+                                       DB_VALUE &boundary_value, bool &clear_boundary_value, const INT16 slot)
+{
+  RECDES left_rec;
+  NON_LEAF_REC non_leaf_rec;
+  int offset;
+
+  if (!m_is_inf_right_key)
```
Comment (ribram) [inline src/storage/btree.c]: might help to split into small functions

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/storage/btree.c (3 incorrect indentation(s) found)
```
 33426 	  ASSERT_ERROR ();
 33495 	      ASSERT_ERROR ();
 33522 	  ASSERT_ERROR ();
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/storage/btree.c [hunk 12]
```diff
@@ -26832,7 +26897,8 @@ btree_split_node_and_advance (THREAD_ENTRY * thread_p, BTID_INT * btid_int, DB_V
 
   /* Find and fix the child to advance to. Use write latch if the child is leaf or if it is already known that it will
    * require an update of max key length. */
-  error_code = btree_search_nonleaf_page (thread_p, btid_int, *crt_page, key, &child_slotid, &child_vpid);
+  error_code = btree_search_nonleaf_page (thread_p, btid_int, *crt_page, key, &child_slotid, &child_vpid,
```
Comment (ribram) [inline src/storage/btree.c]: maybe boundaries argument should be passed only for IB_INSERT since it is not cost-free.

**Pull request merged**